### PR TITLE
Fix scaling of waterfall plots on open

### LINKED ISSUE #32951: Waterfall plot limits don't scale on open
**Describe the bug**
When creating a waterfall plot, the axis limits do not automatically rescale to show all of the data when the plot is first opened. The user has to click the "Reset axes limits" toolbar button.

**To Reproduce**

1. Load a workspace.
2. Right click the workspace and select "Plot > Spectrum".
3. Enter, for example, 1-10 in the spectrum numbers box.
4. Change the plot type to "Waterfall" and click OK.

**Expected behavior**
Axes limits should resalce automatically.

**Screenshots**
image

### PR DESCRIPTION
**Pull request by MialLewis**
**Description of work.**

This fixes the issue of waterfall plots not scaling on open. If the axes is waterfall, following adjustment of the lines, the plot function now calls the relim and autoscale functions to readjust the x and y limits.

In addition, the part of the plot function that overplots lines onto existing waterfall axes has been separated out into its own function to reduce complexity of the plot function (flake 8 complexity check failed).

**To test:**

1) Import data (HRP39182 from training course data used).
2) Plot desired spectra with plot type as waterfall.
3) Confirm suitable axis limits/scaling.
3) Overplot additional spectra onto plotted waterfall, confirm correct plotting and suitable axis limits/scaling.

Fixes #32951.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by robertapplin — CHANGES_REQUESTED** [review]
Your changes work and the code is looking good, so nice work! I just have a few comments below.

I imagine it wouldn't be easy to add a unit test for this?

Also can you add a release note to the bug fix section of this file?https://github.com/mantidproject/mantid/blob/main/docs/source/release/v6.3.0/mantidworkbench.rst

File: Framework/PythonInterface/mantid/plots/plotfunctions.py
Comment (MialLewis) [reply Framework/PythonInterface/mantid/plots/plotfunctions.py:226]: Doxygen comment added - I've used other comments in the file as a template so let me know if this fulfils requirements.

Noted re. python type hints.

File: Framework/PythonInterface/mantid/plots/plotfunctions.py
Comment (MialLewis) [reply Framework/PythonInterface/mantid/plots/plotfunctions.py]: Done.

File: Framework/PythonInterface/mantid/plots/plotfunctions.py
Comment (robertapplin) [inline Framework/PythonInterface/mantid/plots/plotfunctions.py:210]: Thanks for making those changes. Just noticed we can probably remove this line, because we have the elif statement above it 
```suggestion
```

**Review by robertapplin — CHANGES_REQUESTED** [review]

File: Framework/PythonInterface/mantid/plots/plotfunctions.py
Comment (robertapplin) [reply Framework/PythonInterface/mantid/plots/plotfunctions.py:210]: We might also need the `elif superplot and not tiled` line to go back to `if superplot and not waterfall and not tiled`, but its worth checking if this is the case

Comment (MialLewis) [reply Framework/PythonInterface/mantid/plots/plotfunctions.py:210]: We still need to adjust the axis if we are not overplotting. If this `if` statement is removed and these lines brought under the scope of the preceding `elif` statement, that won't happen.

Comment (robertapplin) [reply Framework/PythonInterface/mantid/plots/plotfunctions.py:210]: Ah good point 👍

**Comment by robertapplin** [comment]
Your changes look good, so nice work. I'll approve after you've added a release note to here

https://github.com/mantidproject/mantid/blob/main/docs/source/release/v6.3.0/mantidworkbench.rst

**Comment by MialLewis** [comment]
> Your changes look good, so nice work. I'll approve after you've added a release note to here
> 
> https://github.com/mantidproject/mantid/blob/main/docs/source/release/v6.3.0/mantidworkbench.rst

I've added a commit to do this now, as I initially missed this comment. Is there much more painful than waiting for tests to complete for an added release note! :)

**Review by robertapplin — APPROVED** [review]
Looks good to go

**Pull request merged**
